clib/feldspar_array.h

```diff
@@ -65,63 +62,56 @@ static inline struct array* allocArray( struct array* src )
   return src;
 }
 
-/// Resizing an existing array (the struct array must be initialized).
-static inline struct array* resizeArray( struct array* arr, int32_t size, int32_t len )
+/// Resizing an existing array.
+static inline void* resizeArray( void* arr, int32_t size, int32_t len )
 {
-  assert( arr && "array not initialized" );
-  arr->buffer = realloc( arr->buffer, len*size );
-  arr->length = len;
-  return arr;
+  log_3( "resize %p with size %d and len %d\n", arr, size, len );
+  return realloc( arr, len*size );
 }
 
 /// Array (re)initialization for flat arrays.
-static inline struct array* initArray( struct array* src, int32_t size, int32_t len )
+static inline void* initArray( void* arr, int32_t arrLen, int32_t size, int32_t newLen )
 {
-  src = allocArray( src );
-  if( len != src->length ) {
-    src = resizeArray( src, size, len );
+  log_4( "initArray %p with arrlen %d size %d newLen %d\n", arr, arrLen, size, newLen );
+  if( newLen != arrLen ) {
+    arr = resizeArray( arr, size, newLen );
   }
-  return src;
+  return arr;
 }
 
 /// Free a flat array or an array where all the arrays it contains have been free'd already.
 // TODO: Think about arrays escaping from their scope.
-static inline void freeArray( struct array* arr )
+static inline void freeArray( void* arr, int32_t len )
```

Comment:
The length shouldn't be required for freeing an array.

(I have no idea why my comment boxes are shifting from left to right on the screen.)

Reply:
It makes things simpler elsewhere to always pass both the length and the buffer. See "arrayFun".

Reply:
I'm not a huge fan of adding unused constructs to the syntax tree since that makes it harder to debug the generated programs. If arrayFun has to be used, can't addSize have a guard on the function name and drop the length parameter for freeArray?